pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -851,6 +851,10 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(ctx context.Con
 				klog.ErrorS(err, "Kill container failed", "pod", klog.KRef(runningPod.Namespace, runningPod.Name), "podUID", runningPod.ID,
 					"containerName", container.Name, "containerID", container.ID)
 			}
+			// if the container is killed, the container must not be ready for readiness. #129552
+			if gracePeriodOverride != nil {
+				m.readinessManager.Set(container.ID, proberesults.Failure, pod)
```

Comment:
Readiness probes are still working at this point because they are stopped after all containers are stopped:
https://github.com/kubernetes/kubernetes/blob/58704903c5f3616967ea9e71cb173d7a44868aa3/pkg/kubelet/kubelet.go#L2090-L2094

I don't think we should update readiness forcibly before stopping readiness probes.

Reply:
Based on https://github.com/kubernetes/kubernetes/issues/129552#issuecomment-2591099849, you pointed out that the issue is caused by not returning `probe.Failure` when the ExecProbe fails after the container has already terminated, right? Is `probe.Unknown` being returned, causing the container to remain in the Ready state?
https://github.com/kubernetes/kubernetes/blob/v1.32.2/pkg/probe/exec/exec.go#L82

Reply:
Yes, I think the behavior of the ExecProbe is another bug similar to #106682.

If an HTTP probe is configured, a container gets unready though it takes `periodSeconds * failureThreashold`:

```
apiVersion: v1
kind: Pod
metadata:
  name: "b382559131"
  namespace: "default"
spec:
  terminationGracePeriodSeconds: 60
  restartPolicy: Never
  containers:
  - name: c1
    image: nginx
    readinessProbe:
      httpGet:
        path: /
        port: 80
  - name: c2
    image: "k8s.gcr.io/busybox"
    command:
      - sh
      - -c
      - "_term() { while true; do echo \"hello_term_c2\"; sleep 1; done } ; trap _term SIGTERM; while true; do echo \"helloc2\"; sleep 1; done
```

Reply:
If we need to set readiness to `False` immediately after pod termination gets started as requested in the issue, this fix will be necessary. However, I think we should stop readiness probes before overwriting a container readiness.

Reply:
https://github.com/kubernetes/kubernetes/blob/30ea0d13cdeeb501625e90819886881930e67017/staging/src/k8s.io/api/core/v1/types.go#L2587-L2590

failureThreshold defaults to 3, which means there will be a higher delay by default. I hope to immediately set the container's ready to false after killing the container, to avoid traffic access, because even if the ExecProbe bug is fixed, the container's ready status from true to false will still have a higher delay, which may lead to traffic being introduced to an unavailable container.

Reply:
> However, I think we should stop readiness probes before overwriting a container readiness.

I agree that it does feel a bit off, but currently, it’s not possible to stop the probe worker in KubeRuntimeManager. Even if the probe worker for a terminated container keeps running, as long as the container has already terminated, the probe result won't be changed from Failure regardless of whether there's a bug in ExecProbe or not.

Also, a notable change is planned for the behavior of probes in Pods under termination in https://github.com/kubernetes/kubernetes/pull/125710. I'm curious how this change will affect that PR.
/cc @gjkim42